### ISSUE TITLE
Centroid calculation of degenerate polygons is numerically unstable

### ISSUE THREAD

**Issue by tyrasd** [locationtech/jts#193]
From the source code documentation of [`getCentroid`](https://github.com/locationtech/jts/blob/4c4ccd7fe5c758abb48647da60dacf8e4cd7f9df/modules/core/src/main/java/org/locationtech/jts/algorithm/Centroid.java#L123-L124):

> Degenerate geometry are computed using their effective dimension (e.g. areas may degenerate to lines or points)

At least for some degenerate polygons, this doesn't work as expected and produces unexpected results (see example below). As far as I can tell this is because the area calculations are not implemented in a totally numerically stable way, so degenerate polygons (which should have an area of `0`) actually end up with a small "epsilon" remainder in the variable [`areasum2`](https://github.com/locationtech/jts/blob/4c4ccd7fe5c758abb48647da60dacf8e4cd7f9df/modules/core/src/main/java/org/locationtech/jts/algorithm/Centroid.java#L127) caused by floating point rounding errors. Of course, the following centroid calculations don't work properly as expected in these cases.

Here's an example:

* input geometry: `POLYGON ((10.977073899999999 45.3963939, 10.976983299999999 45.3964183, 10.9769597 45.3963749, 10.977076499999999 45.3963433, 10.9769597 45.3963749, 10.976983299999999 45.3964183, 10.977073899999999 45.3963939))`
* value of `areasum2`: `8.271806125530277E-25`
* value of `totalLength`: `5.28457867453174E-4`
* resulting `getCentroid()` coordinate for this geometry: `(10.666666666666666, 0.0)` :boom: 

I believe the [check for degenerate polygon geometries](https://github.com/locationtech/jts/blob/4c4ccd7fe5c758abb48647da60dacf8e4cd7f9df/modules/core/src/main/java/org/locationtech/jts/algorithm/Centroid.java#L127) should maybe compare against a small epsilon value in order to circumvent this issue (also a comparison check against totalLength could make sense).

**Comment by dr-jts** [comment]
Good find.  Of course comparing a numeric result against zero is a well-known cause of numerical robustness errors, and this is a classic example.  Done in haste many years ago...  

Your suggestion of comparing the area/perimeter ratio against epsilon seems like a good idea to improve this.

One might think this would be a solved problem somewhere?

**Comment by dr-jts** [comment]
I wonder if there is a way to rearrange the calculation to make it numerically stable?

**Comment by tyrasd** [comment]
> computing a centroid point for each triangle, and then averaging them in some way

Isn't that basically what the current implementation is doing? :thinking:

I think, for these degenerate polygons, when using an algorithm intended for regular areas, you'll by definition end up with a point that can be *anywhere* since you're basically at some point dividinging zero by zero for the weighted average.

> I wonder if there is a way to rearrange the calculation to make it numerically stable?

I was wondering that as well. But as far as I know (me not being an expert in this field btw), when working with floating point numbers, the best one can do is to minimize such rounding/truncation errors (e.g. by using more clever [summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm)).

**Comment by dr-jts** [comment]
Yes, I was thinking perhaps there was a way to apply something like the Kahan summation trick to the centroid calculation.  Not sure how that can work at the moment.

**Comment by dr-jts** [comment]
One issue with implementing the idea of checking area << perimeter^2 is the choice of factor for the comparison.   It's not clear what that should be to avoid obviously incorrect results.

Another idea is to check if the computed centroid passes some geometric validity checks.  Obvious ones are that it should be within the bounding box of the input. This is fast to check.  A better check is containment in the convex hull of the input, but this is substantially more computation.  

Neither of these are rigorous, but perhaps would give better results than trying to determine an arbitrary comparison factor.

**Comment by tyrasd** [comment]
Here's what I was thinking:

* start with a regular polygon, and make it slimmer and slimmer until it eventually becomes degenerated to a line
* if we had arbitrary precision "real" numbers available, as the polygon gets slimmer, the calculated area centroid would converge towards the "centroid" calculated with a linestring approximation.
* in reality because we're dealing with floating point numbers, the numeric error associated with the area centroid "explodes" in this limit, while the linestring approximation remains stable all the way down
* it should be possible (either by just trial and error or by actually calculating it) to find a reasonable threshold where the numeric errors from the area-centroid calculations outweigh the systematic approximation error one gets from using the linestring-centroid formula for the job. (My guess would be that the suggested `area / perimeter² < epsilon` condition should work for this, but maybe there are better alternatives.)

Using this reasoning, the obtained factor isn't actually *arbitrary*, is it?

A similar gedankenexperiment could also be done the *linestring (or polygon) degenerated to a point* case.

**Comment by dr-jts** [comment]
I agree that your outline does describe the approach needed to introduce an epsilon-based check for deciding when to switch from area-based to line-based computation.  However, I see two issues:

* It's not clear that a single epsilon value will apply equally well to all geometric situations
* How do you determine when "the numeric errors outweigh the linestring approximation error"?

**Comment by jnh5y** [comment]
I'm jumping into the middle on this one.  Would it be possible to calculate an epsilon based on the geometry's size?  (As a second statement, I don't think this blocks a 1.15.0 release.)
